docs/enhacements/20200316-KCP-scale-in.md

```diff
@@ -0,0 +1,164 @@
+
+## Glossary
+
+Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).
+
+## Summary
+
+This proposal outlines a new functionality for Kubeadm Control Plane. This includes scale-in during upgrade of KCP and necessary checks to ensure that KCP does not run single replica in case of scale-in is required.
+
+## Motivation
+
+New Kubeadm Control Plane default implementation does not support scale-in during upgrade. Main motivation for the current design is to ensure upgrading of a control plane with a single replica. This design is also limited by the expectation of automated provisioning mechanism with some level of additional resources to first scale-out the KCP.
+
+These are challenging limitation in baremetal use cases. Extra baremetal machines might not be available and it is not good starting point to expect operators to run their resources idle for possible upgrades. This is extremely vital in all baremetal workflow clusters where all resources are running without any idle machines.   
+
+### Goals
+
+- To support KCP scale-in during upgrade.
+- To make scale-in available via new enableScaleIn label.
+- Enable scale-in KCP one-by-one.
+- Keep current design as a default implementation to ensure upgrading of a control plane with a single replica. 
+
+### Non-Goals
+
+- To add machineDeploymentStrategyType such as rollingUpdate.
+- To support maxUnavailable or maxSurge
+
+## Proposal
+
+### User Stories
+
+#### Story 1 
+
+As an operator I would like to upgrade my baremetal control plane machines without having additional resources for scale-out first.
+
+#### Story 2
+
+As an operator I want control plane provider to make sure that my control plane is not upgraded if KCP is running a single replica and extra resources are not available.
+
+#### Identified Features
+
+1. Control plane provider must be able to scale-in the number of replicas of a control plane from 3 replicas and above, meeting user story 1.
+2. Control plane provider should fall into default implementation when extra resources are available.
+3. Upgrade should fail when KCP is running a single replica and extra resources are not available, meeting user story 2.
+
+### Implementation Details
+
+### New API changes
+
+KubeadmControlPlane:
+
+```
+// KubeadmControlPlaneSpec defines the desired state of KubeadmControlPlane.
+type KubeadmControlPlaneSpec struct {
+	// Number of desired machines. Defaults to 1. When stacked etcd is used only
+	// odd numbers are permitted, as per [etcd best practice](https://etcd.io/docs/v3.3.12/faq/#why-an-odd-number-of-cluster-members).
+	// This is a pointer to distinguish between explicit zero and not specified.
+	// +optional
+	Replicas *int32 `json:"replicas,omitempty"`
+
+	// EnableScaleIn is a field to indicate that deployment can be scale in during update.
+	// EnableScaleIn can be performed only if replica count is > 1.
+	// +optional
+	EnableScaleIn bool `json:"enableScaleIn"`
```

Comment:
Instead of having a boolean field that enables/disables certain behavior, could we instead opt for something similar to MachineDeployment? https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#proportional-scaling

Reply:
@vincepri I thing we should. Having full features of MachineDeployment in KCP would make life easier also in huge baremetal clusters etc. I support idea of having maxSurge and maxUnavailability in KCP.